go/engine/crypto.go

```diff
@@ -116,8 +149,163 @@ func UnboxBytes32(g *libkb.GlobalContext, secretUI libkb.SecretUI,
 
 	copy(bytes32[:], decryptedData)
 	return
+
 }
 
-func UnboxBytes32Any(g *libkb.GlobalContext, secretUI libkb.SecretUI, arg keybase1.UnboxBytes32AnyArg) (res keybase1.UnboxAnyRes, err error) {
-	return
+func getMatchingSecretKey(g *libkb.GlobalContext, secretUI libkb.SecretUI, arg keybase1.UnboxBytes32AnyArg) (key libkb.GenericKey, index int, err error) {
+	// first check cached keys
+	key, index, err = matchingCachedKey(g, arg)
+	if err != nil {
+		return nil, 0, err
+	}
+	if key != nil {
+		return key, index, nil
+	}
+
+	// load the user
+	me, err := libkb.LoadMe(libkb.NewLoadUserArg(g))
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// check the device key for this user
+	key, index, err = matchingDeviceKey(g, secretUI, arg, me)
+	if err != nil {
+		return nil, 0, err
+	}
+	if key != nil {
+		return key, index, nil
+	}
+
+	// check the paper keys for this user
+	key, index, err = matchingPaperKey(g, secretUI, arg, me)
+	if err != nil {
+		return nil, 0, err
+	}
+	if key != nil {
+		return key, index, nil
+	}
+
+	return nil, 0, libkb.NoSecretKeyError{}
+}
+
+// check cached keys for arg.Pairs match.
+func matchingCachedKey(g *libkb.GlobalContext, arg keybase1.UnboxBytes32AnyArg) (key libkb.GenericKey, index int, err error) {
+	err = g.LoginState().Account(func(a *libkb.Account) {
+		// check device key first
+		dkey, err := a.CachedSecretKey(libkb.SecretKeyArg{KeyType: libkb.DeviceEncryptionKeyType})
+		if err == nil {
+			if n, ok := kidMatch(dkey, arg.Pairs); ok {
```

Comment:
can you factor out the repeated code?  And should some of this logic be moved into `Account`?